File: ucloud-ufile-sdk/OC/UFileSDKDemo/UFileSDKDemo/MultipartUploadViewController.m
//
//  MultipartUploadViewController.m
//  UFileSDKDemo
//
//  Created by 涂雄 on 2017/11/7.
//  Copyright © 2017年 涂雄. All rights reserved.
//

#import "MultipartUploadViewController.h"
#import "MultipartTableViewController.h"
@interface MultipartUploadViewController ()
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIButton *btnSartUpload;
@property (weak, nonatomic) IBOutlet UIButton *btnFinishUpload;
@property (weak, nonatomic) IBOutlet UIButton *btnCancleUpload;
@property (nonatomic, strong) MultipartTableViewController* tableController;

@property (nonatomic, strong) MultipartSession* session;
@end

@implementation MultipartUploadViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.tableController = [[MultipartTableViewController alloc] init];
    self.tableView.delegate = self.tableController;
    self.tableView.dataSource = self.tableController;
    [self.tableView reloadData];
    self.btnSartUpload.enabled = YES;
    self.btnFinishUpload.enabled = NO;
    self.btnCancleUpload.enabled = NO;
   
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

- (IBAction)startToPart:(id)sender {
    
    __weak typeof(self) weakself = self;
    self.btnSartUpload.enabled = NO;
    self.btnFinishUpload.enabled = YES;
    self.btnCancleUpload.enabled = YES;
//    NSString* strPath = [[NSBundle mainBundle] pathForResource:@"test" ofType:@"mp4"];
    NSString* strPath = [[NSBundle mainBundle] pathForResource:@"initscreen" ofType:@"jpg"];
    if (nil == strPath) {
        UIAlertController* alterView =  [UIAlertController alertControllerWithTitle:@"Warning!" message:@"路径不存在" preferredStyle:(UIAlertControllerStyleAlert)];
        UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [alterView addAction:okAction];
        [self presentViewController:alterView animated:YES completion:nil];
    }
   
//    NSString* strPath = [[[NSBundle mainBundle] bundlePath]stringByAppendingPathComponent:@"test.mp4"];
    NSString* strkey = @"initscreen.jpg";
    NSString* contentType = @"image/jpeg";
    NSString* strAuth = [self.ufileSDK calcKey:@"POST" Key:strkey MD5:@"" ContentType:contentType CallBackPolicy:nil];
    [self.ufileSDK.ufileApi multipartUploadStart:strkey authorization:strAuth fileType:contentType
                                         success:^(NSDictionary * _Nonnull response) {
                                             if (response) {
                                                 NSString* msg = [NSString stringWithFormat:@"Multipart start:\n uploadId=%@\n blocksize=%lu \n bucketName=%@ \n Key=%@\n",response[kUFileRespUploadId],[response[kUFileRespBlockSize] integerValue],response[kUFileRespBucketName],response[kUFileRespKeyName]];
                                                 
                                                 weakself.session = [[MultipartSession alloc] init:self.ufileSDK UPLoadID:response[kUFileRespUploadId] BlockSize:[response[kUFileRespBlockSize] integerValue] FileURL:strPath KEY:strkey];
                                                 weakself.session = [[MultipartSession alloc] init:self.ufileSDK UPLoadID:response[kUFileRespUploadId] BlockSize:[response[kUFileRespBlockSize] integerValue] FileURL:strPath KEY:strkey];
                                                 [[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"PartSucess"];
                                                 [weakself.tableController setSession:self.session];
                                                 [weakself.tableView reloadData];
                                                 [weakself showMsg:msg];
                                             }
                                             
                                             
                                         } failure:^(NSError * _Nonnull error) {
                                             
                                             [self showError:error];
                                         }];
}

- (IBAction)finishUpload:(id)sender {
    
    __weak typeof(self) weakself = self;
    self.btnCancleUpload.enabled = NO;
    self.btnFinishUpload.enabled = NO;
    self.btnSartUpload.enabled = YES;
    NSDictionary* policyDic = [NSDictionary dictionaryWithObjectsAndKeys:
                               @"http://test.ucloud.cn",@"callbackUrl",
                               @"url=http://demo.ufile.ucloud.cn/test.mp4&patten_name=mypolicy",
                               @"callbackBody",nil];
    NSString* contentType = @"image/jpeg";
    NSString* strAuth = [self.ufileSDK calcKey:@"POST" Key:self.session.key MD5:@"" ContentType:contentType CallBackPolicy:policyDic];
    [self.ufileSDK.ufileApi multipartUploadFinish:self.session.key uploadId:self.session.uploadID newKey:[NSString stringWithFormat:@"%@%@",self.session.key,@""] etags:self.session.etags contentType:contentType authorization:strAuth
    success:^(NSDictionary * _Nonnull response) {
        
        NSString* msg = [NSString stringWithFormat:@"File Upload Finshed:\n  blocksize=%lu \n bucketName=%@ \n Key=%@\n",response[kUFileRespBlockSize],response[kUFileRespBucketName],response[kUFileRespKeyName]];
        [weakself showMsg: msg];
        
    } failure:^(NSError * _Nonnull error) {
        [weakself showError:error];
    }];
    
}
- (IBAction)cancelUpload:(id)sender {
    
    __weak typeof(self) weakself = self;
    self.btnCancleUpload.enabled = NO;
    self.btnFinishUpload.enabled = NO;
    self.btnSartUpload.enabled = YES;
    NSString* contentType = @"image/jpeg";
    NSString* strAuth = [self.ufileSDK calcKey:@"DELETE" Key:self.session.key MD5:@"" ContentType:contentType CallBackPolicy:nil];
    
    [self.ufileSDK.ufileApi multipartUploadAbort:self.session.key uploadId:self.session.uploadID  authorization:strAuth contentType:contentType success:^(NSDictionary * _Nonnull response) {
                                              
                                              NSString* msg = @"File Upload aborted!";
                                              [weakself showMsg: msg];
                                              
                                          } failure:^(NSError * _Nonnull error) {
                                              [weakself showError:error];
                                          }];
}

-(void)showMsg:(NSString*)msg
{
    UIAlertController* alterView =  [UIAlertController alertControllerWithTitle:@"File upload  Fininshed!" message:msg preferredStyle:(UIAlertControllerStyleAlert)];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alterView addAction:okAction];
    [self presentViewController:alterView animated:YES completion:nil];
}

-(void)showError:(NSError*)error
{
    NSString* errMsg = nil;
    if (error.domain == kUFileSDKAPIErrorDomain) {
        
        errMsg = error.userInfo[@"ErrMsg"];
    }
    else
    {
        errMsg = error.description;
    }
        
    UIAlertController* alterView =  [UIAlertController alertControllerWithTitle:@"File upload  Failed!" message:errMsg preferredStyle:(UIAlertControllerStyleAlert)];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
    [alterView addAction:okAction];
    [self presentViewController:alterView animated:YES completion:nil];
    
}

@end

File: ucloud-ufile-sdk/Example/ucloud-ufile-sdk/MultipartUploadViewController.swift
//
//  MultipartUploadViewController.swift
//  ufile sdk demo
//
//  Created by wu shauk on 12/16/15.
//  Copyright © 2015 ucloud. All rights reserved.
//

import UIKit
import ucloud_ufile_sdk

class MultipartUploadViewController : UIViewController {
    
    @IBOutlet weak var table: UITableView!
    @IBOutlet weak var startUpload: UIButton!
    
    @IBOutlet weak var cancelBtn: UIButton!
    
    @IBOutlet weak var finishUploadBtn: UIButton!
    
    var ufileSDK : UFileSDK?
    var session: MultipartSession?
    var tableController: MultipartTableController?
    
    @IBOutlet weak var upload: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableController = MultipartTableController()
        self.table.delegate = self.tableController
        self.table.dataSource = self.tableController
        self.table.reloadData()
        self.startUpload.isEnabled = true
        self.finishUploadBtn.isEnabled = false
        self.cancelBtn.isEnabled = false
    }
    
    @IBAction func uploadBtnClicked(_ sender: AnyObject) {
        
        self.cancelBtn.isEnabled = true
        self.finishUploadBtn.isEnabled = true
        self.startUpload.isEnabled = false

        let b = Bundle.main.path(forResource: "initscreen", ofType: "jpg")
        let key = "test-multipart.jpg"
 //       let data = NSData(contentsOfFile: b!)
        let auth = ufileSDK!.calcKey("POST", key: key, contentMd5: nil, contentType: nil)
        ufileSDK!.ufileSDK!.multipartUploadStart(key, authorization: auth,
            success: { (result: [AnyHashable: Any]) -> Void in
                let msg = String(format: "Multipart start:\n uploadId=%@\nblockSize=%lu\nbucketName=%@\nKey=%@\n", arguments: [result[kUFileRespUploadId] as! NSString, (result[kUFileRespBlockSize] as! NSNumber).intValue, result[kUFileRespBucketName] as! NSString, result[kUFileRespKeyName] as! NSString]);
                self.session = MultipartSession(info: self.ufileSDK!, uploadId: result[kUFileRespUploadId] as! NSString as String, blockSize: UInt(result[kUFileRespBlockSize] as! NSNumber), fileURL: b!, key: key)
                self.tableController?.setSession(self.session!)
                self.table.reloadData()
                self.showMsg(msg)
                
            },
            failure: { (err) -> Void in
                self.showError(err as NSError)
        })
    }
    
    fileprivate func showMsg(_ msg: String) {
        let alert = UIAlertController(title: "File Upload Done", message: msg, preferredStyle: UIAlertControllerStyle.alert)
        let okAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.default) {
            UIAlertAction in
        }
        alert.addAction(okAction)
        self.present(alert, animated: true, completion: nil)
    }
    
    fileprivate func showError(_ err: NSError) {
        var errMsg: String? = nil;
        if err.domain == kUFileSDKAPIErrorDomain {
            errMsg = err.userInfo["ErrMsg"] as? NSString as String?;
            NSLog("%@", err.userInfo);
        } else {
            errMsg = err.description
        }
        let alert = UIAlertController(title: "File Upload Fail", message: errMsg, preferredStyle: UIAlertControllerStyle.alert)
        let okAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.default) {
            UIAlertAction in
        }
        alert.addAction(okAction)
        self.present(alert, animated: true, completion: nil)
    }
    
    @IBAction func finishUpload(_ sender: AnyObject) {
        self.cancelBtn.isEnabled = false
        self.finishUploadBtn.isEnabled = false
        self.startUpload.isEnabled = true
        let auth = self.ufileSDK!.calcKey("POST", key: self.session!.key, contentMd5: nil, contentType: "application/jpeg")
        self.ufileSDK!.ufileSDK!.multipartUploadFinish(self.session!.key, uploadId: self.session!.uploadId, newKey: self.session!.key+NSTimeIntervalSince1970.description, etags: self.session!.etags, contentType: "application/jpeg", authorization: auth, success: { (result:[AnyHashable: Any]) -> Void in
                let s = String(format: "File upload success\nBucket=%@\nKey=%@\nFileSize=%@", arguments: [result[kUFileRespBucketName] as! NSString, result[kUFileRespKeyName] as! NSString, result[kUFileRespLength] as! NSNumber])
                self.showMsg(s)
            }) { (err) -> Void in
                self.showError(err as NSError)
        }
        
        self.session = nil
    }
    
    @IBAction func cancelUpload(_ sender: AnyObject) {
        self.cancelBtn.isEnabled = false
        self.finishUploadBtn.isEnabled = false
        self.startUpload.isEnabled = true
        let auth = self.ufileSDK!.calcKey("DELETE", key: self.session!.key, contentMd5: nil, contentType: nil)
        self.ufileSDK!.ufileSDK!.multipartUploadAbort(self.session!.key, uploadId: self.session!.uploadId, authorization: auth,success: { (result:[AnyHashable: Any]) -> Void in
            let s = String("File upload aborted")
            self.showMsg(s!)
            }) { (err) -> Void in
                self.showError(err as NSError)
        }
        self.session = nil

    }
    
}

